core/field_textinput.js

```diff
@@ -365,12 +404,77 @@ Blockly.FieldTextInput.prototype.onHtmlInputChange_ = function(_e) {
     Blockly.Events.setGroup(true);
     this.setValue(text);
     // Always render the input text.
-    this.text_ = this.htmlInput_.value;
+
+    // As I mentioned in the doValueUpdate_ function, the onHtmlInputChange
+    // handler deals with setting the display value.
+    // This makes sure we get the input "straight from the horses mouth"
+    // without any validators changing, or invalidating the input.
+
+    // This also means we have no need to override the doValueInvalid_
+    // function. A nice bonus.
+    this.displayValue_ = this.htmlInput_.value;
     this.forceRerender();
     Blockly.Events.setGroup(false);
   }
 };
 
+// Now you said that you did not want a solution that just renames text_.
+// Let me go ahead and speak to that for a moment.
+
+// Most of the problems with text_ have already been fixed prior to this PR.
+//   1) Setting the text_ property (for use in collapse text) was inefficient.
+//   This has been solved by moving people to the getText property.
+//   2) Moving peoples code to getText keeps it more organized, instead of
+//   having it split between doValueUpdate, doValueInvalid_ and render_.
+// As you know I also noted these facts in my first comment:
+// https://github.com/google/blockly/issues/2720#issuecomment-515688519
+
```

Comment:
The third reason for getting rid of text_ is to make it clear to developers this is an intentional change in behavior and usage. We think using text_ as a way to override what's being displayed was broken previously and we are now providing an explicit way to do this override.